cmd/segment-reaper/detect_test.go

```diff
@@ -0,0 +1,635 @@
+					if inline {
+						expectedInlineSegments++
+						expectedRemoteSegments += (numSegments - 1)
+					} else {
+						expectedRemoteSegments += numSegments
+					}
+				}
+			}
+
+			ctx := testcontext.New(t)
+			defer ctx.Cleanup()
+
+			for _, objSeg := range objSegments {
+				err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+				require.NoError(t, err)
+			}
+
+			assert.Equal(t, projID.String(), obsvr.lastProjectID, "lastProjectID")
+			assert.Equal(t, expectedInlineSegments, obsvr.inlineSegments, "inlineSegments")
+			// newObject if returns an inline segment is always the last
+			assert.Equal(t, expectedInlineSegments, obsvr.lastInlineSegments, "lastInlineSegments")
+			assert.Equal(t, expectedRemoteSegments, obsvr.remoteSegments, "remoteSegments")
+
+			if assert.Equal(t, len(expectedObjects), len(obsvr.objects), "objects number") {
+				for cluster, bucketObjs := range obsvr.objects {
+					expBucketObjs, ok := expectedObjects[cluster.bucket]
+					if !ok {
+						t.Errorf("bucket '%s' shouldn't exist in objects map", cluster.bucket)
+						continue
+					}
+
+					if !assert.Equalf(t, len(expBucketObjs), len(bucketObjs), "objects per bucket (%s) number", cluster.bucket) {
+						continue
+					}
+
+					for expPath, expObj := range expBucketObjs {
+						if !assert.Contains(t, bucketObjs, expPath, "path in bucket objects map") {
+							continue
+						}
+
+						obj := bucketObjs[expPath]
+						assert.Equal(t, expObj.expectedNumberOfSegments, obj.expectedNumberOfSegments, "Object.expectedNumSegments")
+						assert.Equal(t, expObj.hasLastSegment, obj.hasLastSegment, "Object.hasLastSegment")
+
+						assert.Equal(t, expObj.skip, obj.skip, "Object.skip")
+						if !expObj.skip {
+							assert.Equal(t, expObj.segments, obj.segments, "Object.segments")
+						}
+					}
+				}
+			}
+		})
+	})
+
+	t.Run("analyzeProject", func(t *testing.T) {
+		t.Run("object without last segment", func(t *testing.T) {
+			var objectsMap ObjectsMap
+			{ // Generate an objects without last segment
+				const (
+					bucketName = "analyzeBucket"
+					objPath    = storj.Path("analyzePath")
+				)
+				var segments uint64
+				{
+					numSegments := rand.Intn(62) + 1
+					segments = math.MaxUint64 >> numSegments
+				}
+
+				projID, err := uuid.New()
+				require.NoError(t, err)
+				objectsMap = ObjectsMap{
+					Cluster{
+						projectID: projID.String(),
+						bucket:    bucketName,
+					}: map[storj.Path]*Object{
+						objPath: {
+							segments:       segments,
+							hasLastSegment: false,
+						},
+					},
+				}
+			}
+
+			var (
+				buf   = &bytes.Buffer{}
+				obsvr = Observer{
+					db:      teststore.New(),
+					writer:  csv.NewWriter(buf),
+					objects: objectsMap,
+				}
+			)
+
+			ctx := testcontext.New(t)
+			defer ctx.Cleanup()
+
+			err := analyzeProject(ctx.Context, obsvr.db, obsvr.objects, obsvr.writer)
+			require.NoError(t, err)
+
+			// TODO: Add assertions for buf content
```

Comment:
is this TODO to finish in this PR or another one?

Reply:
Yes, this branch doesn't contain the `analyzeProject` logic so I didn't know what to expect, however if you think that's better to add them in this branch, then I can take a look to the implementation in https://github.com/storj/storj/pull/3598 and add them.

Let me know if you think that's better.

Reply:
sounds good to do this with another PR